### ISSUE TITLE
await doesn't always wait

### ISSUE THREAD

**Issue by hpmv** [dart-lang/sdk#29906]
The [async await page](https://www.dartlang.org/articles/language/await-async) of Dart mentions that "await always waits". This means that the following code has predictable behavior:

```
main() async {
  scheduleMicrotask(() => print(1));
  await 0;
  print(2);
}
```

will print 1 and then 2.

However, I can break this semantics by carefully constructing some futures. Here we go:
```
import 'dart:async';

Future m2(Future f) async {
  await f;
}

Future m1(Future f) async {
  await m2(f);
}

Future normalCase() async {
  print('awaiting');
  scheduleMicrotask(() => print(1));
  await 0;
  scheduleMicrotask(() => print(2));
  await 0;
  scheduleMicrotask(() => print(3));
  await 0;
  scheduleMicrotask(() => print(4));
  await 0;
  scheduleMicrotask(() => print(5));
  await 0;
}

Future strangeCase() async {
  final f = new Future.delayed(new Duration(microseconds: 100));
  final f1 = m1(f);
  final f2 = m1(f);
  final f3 = m1(f);
  final f4 = m1(f);
  final f5 = m1(f);
  print('awaiting');
  scheduleMicrotask(() => print(1));
  await f1;
  scheduleMicrotask(() => print(2));
  await f2;
  scheduleMicrotask(() => print(3));
  await f3;
  scheduleMicrotask(() => print(4));
  await f4;
  scheduleMicrotask(() => print(5));
  await f5;
}

main() async {
  print('Running normal case');
  await normalCase();
  print('Normal case done');
  print('');
  
  await new Future.delayed(new Duration(microseconds:100));
  
  print('Running strange case');
  await strangeCase();
  print('strange case done');
}

```

Output:
```

HTML OUTPUT
CONSOLE
Running normal case
awaiting
1
2
3
4
5
Normal case done

Running strange case
awaiting
1
strange case done
2
3
4
5
```

In the strange case, the completion of the Future `f` synchronously completes all five awaits as well as the await in main(), which violates the "await always waits" property.

I think the reason this happens is:

1. 'await' synchronously registers a listener on the future being awaited on (nothing surprising).
2. An async function seems to be implemented as a _SyncCompleter, so when the async function returns, it synchronously completes any awaiters. (This itself is fine since we've already paused when executing 'await').
3. When a future being await-ed on completes, the code after the await expression is executed synchronously, on top of the call stack, inside the _propagateToListeners call. (This itself is also fine).
4. _propagateToListeners take the current set of listeners, and calls them one by one. (correct behavior of course).

However combining them leads to the following:

* Since we made `f` take a long time to complete, before it completes, five instances of `m2` are awaiting on `f`, and five instances of `m1` are awaiting on the corresponding `m2`. So `f` has five listeners.
* Future `f` completes, and invokes the five listeners one by one:
  * The first listener is invoked, continuing first instance of `m2`, which synchronously completes its async function future, which synchronously continues the first instance of `m1`, which synchronously continues the code after `await f1`.
  * `await f2` registers a new listener on the second instance of `m1`.
  * The first listener returns. The second listener is invoked, continuing the second instance of `m2`, which continues the second instance of `m1`, which continues the code after `await f2`.
  * `await f3` registers a new listener on the third instance of `m2`.
  * etc. until all five awaits are synchronously completed.

This does not seem like intended behavior to me, especially since the behavior becomes normal if we just slightly alter the setup, such as awaiting in the wrong order (f2, f1, f3, f4, f5).

I think the root cause is that we're copying the future listeners shallowly before invoking them, so more listeners can sneak in while we invoke the top-level ones. But it sounds non-trivial to deep-copy them. However I'm not sure if we want to change any of 1 to 3 above, as they all seem to have good reasons. Someone with deeper knowledge of Dart can advise.

**Comment by lrhn** [comment]
In short: Future completion isn't always using the microtask queue, sometimes it just cascades completion from one future to the next as they happen. This is a deliberate choice, and it affects `await` because `await` is just a syntactic shorthand for calling `Future.then`.

The `await` promises to delay execution until *a* later microtask, but it doesn't promise that it waits until after *all* currently scheduled microtasks. We could have chosen to do that, but in most cases, it would just introduce unnecessary overhead, and it wouldn't match the behavior you get from writing future code manually with `then` calls.
Instead, the `await` delays until the future completes. That may be seconds later, or almost right away. We do promise that that happens in a later microtask, so there will be a slight delay, but the future completion might already have been scheduled and it can happen as early as the very next microtask. If so, the `await` will continue executing at that point, way before the `scheduleMicrotask` on the line before, which does promise to go at the end of the microtask queue.

More precisely, if you are awaiting for a future to complete, the code after the `await` will run immediately when the future completes. It will not go at the end of the microtask queue, it will run then-and-there and if that happens to complete an async function, the future returned by that async function is also completed immediately, and the future completion can cascade until all depending futures are handled.
That works just the same as a chain of futures `future.then(...).then(...).then(...)`. If all the callbacks are synchronous, this all completes when the first future completes.

You see a difference when you swap the futures because you then await `f1` after `f2`, but `f1` completes (just) before `f2`. That means that you wait for `f1` *after* it has already completed. If that happens, it can't do something immediately (future completion mustn't happen in the middle of other code), so it has to wait for a later microtask. It schedules itself (using `scheduleMicrotask`) to tell the listener that it is complete. That puts it at the end of the microtask queue (we could put it at the beginning instead, but we haven't seen any reason for doing that).

So, in summary:
* `await f` waits for the future to complete, then continues when that happens (just as calling `f.then`, not coincidentally).
* Future completion doesn't use the microtask queue, they cascade their completion to their listeners immediately.
* If the future is already complete, it has to schedule itself to repeat the result later, so it does goes on the microtask stack.
* An async function that terminates will complete its future at that point.
* Don't depend on an `await` delaying until after all previously scheduled microtasks, because often it won't.

That is why:
```dart
import "dart:async";
foo() async {
  scheduleMicrotask(() => print("mt-foo1"));
  print(">foo");
  await bar();
  print("<foo");
  scheduleMicrotask(() => print("mt-foo2"));
}
bar() async {
  scheduleMicrotask(() => print("mt-bar1"));
  print(">bar");
  await baz();
  print("<bar");
  scheduleMicrotask(() => print("mt-bar2"));
}
baz() async {
  scheduleMicrotask(() => print("mt-baz"));
  print("baz");
}
main() async {
  print("start");
  scheduleMicrotask(() => print("mt-main1"));
  await foo();
  scheduleMicrotask(() => print("mt-main2"));
  print("done");
}
```
prints
```
start
mt-main1
>foo
mt-foo1
>bar
mt-bar1
baz
<bar
<foo
done
mt-baz
mt-bar2
mt-foo2
mt-main2
```
You can see that `baz`, `<bar`, `<foo`, and `done` come immediately after each other because they form a future completion cascade. Then the microtasks scheduled along that path will trigger afterwards.

**Comment by hpmv** [comment]
Thanks for the explanation! I did not know that microtasks might go out of order in Dart in anyway. (Unfortunately then, `await` isn't as predictable as I expected.)

One thing that still bothers me is that technically, the await f2, await f3, await f4, and await f5 didn't happen in microtasks at all, but all within the _propagateToListeners call. I'm trying to convince myself that this is equivalent to scheduling these awaits at the beginning of the microtask queue (in other words, the listeners can be seen as the queue) but that would rely on the assumption that the completion of `f` is at the bottom of the call stack (i.e. is the only thing that happens before yielding control to the zone). Or in general, synchronous future completion must only be triggerable by a microtask, a timer, or another future completion, but not user code (except when that code returns).

In other words, is the following true in general, that the user cannot ever do something like this:

```
// run some code here
// do something here which synchronously completes a future
// run some more code here
```

because if I could do that, then the second part of "run some more code here" would be executed after those awaits, which would make the awaits not appear to have been executed in any microtasks.

**Comment by floitschG** [comment]
There are no plans to change the behavior.

**Comment by santosh81066** [comment]
hello actually my function is not awaiting to complete process and jumping to next code
